discovery_engine/lib/src/domain/search_manager.dart

```diff
@@ -0,0 +1,217 @@
+    await _engineStateRepo.save(await _engine.serialize());
+
+    await _docRepo.updateMany(searchDocs.map((e) => e.document));
+    for (final docWithData in searchDocs) {
+      final id = docWithData.document.documentId;
+      await _activeRepo.update(id, docWithData.data);
+    }
+
+    return searchDocs
+        .map((docWithData) => docWithData.document.toApiDocument())
+        .toList();
+  }
+
+  /// Obtain the first batch of search documents and persist to repositories.
+  Future<EngineEvent> searchRequested(
+    String queryTerm,
+    FeedMarket market,
+  ) async {
+    await searchClosed();
+
+    final search = ActiveSearch(
+      queryTerm: queryTerm,
+      requestedPageNb: 1,
+      pageSize: _config.maxSearchDocs,
+      market: market,
+    );
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.searchRequestFailed(reason);
+    }
+
+    // TODO: should we persist active search before we get documents?
+    await _searchRepo.save(search);
+
+    return EngineEvent.searchRequestSucceeded(search, docs);
+  }
+
+  /// Obtain the next batch of search documents and persist to repositories.
+  Future<EngineEvent> nextSearchBatchRequested() async {
+    var search = await _searchRepo.getCurrent();
+
+    if (search == null) {
+      const reason = SearchFailureReason.noActiveSearch;
+      return const EngineEvent.nextSearchBatchRequestFailed(reason);
+    }
+
+    // lets update active search params
+    search = search.copyWith(requestedPageNb: search.requestedPageNb + 1);
+    final docs = await _getSearchDocuments(search);
+
+    if (docs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.nextSearchBatchRequestFailed(reason);
+    }
+
+    // TODO: should we persist updated search params before we get documents?
+    await _searchRepo.save(search);
+
+    return EngineEvent.nextSearchBatchRequestSucceeded(search, docs);
+  }
+
+  /// Returns the list of active search documents, ordered by their global rank.
+  ///
+  /// That is, documents are ordered by their timestamp, then local rank.
+  Future<EngineEvent> restoreSearchRequested() async {
+    final search = await _searchRepo.getCurrent();
+
+    if (search == null) {
+      const reason = SearchFailureReason.noActiveSearch;
+      return const EngineEvent.restoreSearchFailed(reason);
+    }
+
+    final allDocs = await _docRepo.fetchAll();
+    final searchDocs = allDocs
+        // we only want active search documents
+        .where((doc) => doc.isSearched && doc.isActive)
+        .toList();
+
+    if (searchDocs.isEmpty) {
+      const reason = SearchFailureReason.noResultsAvailable;
+      return const EngineEvent.restoreSearchFailed(reason);
+    }
+
+    searchDocs.sort((doc1, doc2) {
+      final timeOrd = doc1.timestamp.compareTo(doc2.timestamp);
```

Comment:
this sorts older timestamps to the start and newer timestamps to the end, is this intended or do we want to have newer docs first? then we should revert the order.
```suggestion
      final timeOrd = doc2.timestamp.compareTo(doc1.timestamp);
```

Reply:
good question, but I don't know the answer, maybe @acrrd can shed some light on it?

Reply:
older first because we want to return them in the same order user saw them the first time. This should be the same as we are doing for the feed